### PR TITLE
[14.0][MIG] l10n_it_withholding_tax

### PR DESCRIPTION
**Pull request by TheMule71**
Porting di l10n_it_withholding_tax dalla 12.0 alla 14.0

DA COMPLETARE

--
Confermo di aver firmato il CLA https://odoo-community.org/page/cla e di aver letto le linee guida su https://odoo-community.org/page/contributing

### CODE REVIEW

**Comment by eLBati** [comment]
A proposito della riconciliazione: il widget che usavamo è stato spostato in enterprise; in OCA viene reintrodotto tramite https://github.com/OCA/account-reconcile/pull/359

Sarà probabilmente necessario creare 2 moduli "ponte"
 - uno che dipende da ritenute + `account_reconciliation_widget`
 - uno che dipende da ritenute + il modulo enterprise

**Comment by CiroBoxHub** [comment]
Ciao, volevo segnalare questo errore che si verifica al momento della scelta di un prodotto in una nuova riga fattura quando si sta creando una nuova fattura.
Non appena si seleziona un prodotto appare l'errore:
`Errore:
Odoo Server Error

Traceback (most recent call last):
  File "/home/cirobox/odoo/odoo14/apps/odoo/lib/odoo-14.0.post20210130-py3.7.egg/odoo/addons/base/models/ir_http.py", line 237, in _dispatch
    result = request.dispatch()
  File "/home/cirobox/odoo/odoo14/apps/odoo/lib/odoo-14.0.post20210130-py3.7.egg/odoo/http.py", line 683, in dispatch
    result = self._call_function(**self.params)
  File "/home/cirobox/odoo/odoo14/apps/odoo/lib/odoo-14.0.post20210130-py3.7.egg/odoo/http.py", line 359, in _call_function
    return checked_call(self.db, *args, **kwargs)
  File "/home/cirobox/odoo/odoo14/apps/odoo/lib/odoo-14.0.post20210130-py3.7.egg/odoo/service/model.py", line 94, in wrapper
    return f(dbname, *args, **kwargs)
  File "/home/cirobox/odoo/odoo14/apps/odoo/lib/odoo-14.0.post20210130-py3.7.egg/odoo/http.py", line 347, in checked_call
    result = self.endpoint(*a, **kw)
  File "/home/cirobox/odoo/odoo14/apps/odoo/lib/odoo-14.0.post20210130-py3.7.egg/odoo/http.py", line 912, in __call__
    return self.method(*args, **kw)
  File "/home/cirobox/odoo/odoo14/apps/odoo/lib/odoo-14.0.post20210130-py3.7.egg/odoo/http.py", line 531, in response_wrap
    response = f(*args, **kw)
  File "/home/cirobox/odoo/odoo14/apps/odoo/lib/odoo-14.0.post20210130-py3.7.egg/odoo/addons/web/controllers/main.py", line 1377, in call_kw
    return self._call_kw(model, method, args, kwargs)
  File "/home/cirobox/odoo/odoo14/apps/odoo/lib/odoo-14.0.post20210130-py3.7.egg/odoo/addons/web/controllers/main.py", line 1369, in _call_kw
    return call_kw(request.env[model], method, args, kwargs)
  File "/home/cirobox/odoo/odoo14/apps/odoo/lib/odoo-14.0.post20210130-py3.7.egg/odoo/api.py", line 396, in call_kw
    result = _call_kw_multi(method, model, args, kwargs)
  File "/home/cirobox/odoo/odoo14/apps/odoo/lib/odoo-14.0.post20210130-py3.7.egg/odoo/api.py", line 383, in _call_kw_multi
    result = method(recs, *args, **kwargs)
  File "/home/cirobox/odoo/odoo14/apps/odoo/lib/odoo-14.0.post20210130-py3.7.egg/odoo/addons/account/models/account_move.py", line 1048, in onchange
    return super(AccountMove, self.with_context(recursive_onchanges=False)).onchange(values, field_name, field_onchange)
  File "/home/cirobox/odoo/odoo14/apps/odoo/lib/odoo-14.0.post20210130-py3.7.egg/odoo/models.py", line 6234, in onchange
    for name in nametree
  File "/home/cirobox/odoo/odoo14/apps/odoo/lib/odoo-14.0.post20210130-py3.7.egg/odoo/models.py", line 6235, in <listcomp>
    if name not in done and snapshot0.has_changed(name)
  File "/home/cirobox/odoo/odoo14/apps/odoo/lib/odoo-14.0.post20210130-py3.7.egg/odoo/models.py", line 6066, in has_changed
    return self[name] != record[name]
  File "/home/cirobox/odoo/odoo14/apps/odoo/lib/odoo-14.0.post20210130-py3.7.egg/odoo/models.py", line 5664, in __getitem__
    return self._fields[key].__get__(self, type(self))
  File "/home/cirobox/odoo/odoo14/apps/odoo/lib/odoo-14.0.post20210130-py3.7.egg/odoo/fields.py", line 934, in __get__
    self.recompute(record)
  File "/home/cirobox/odoo/odoo14/apps/odoo/lib/odoo-14.0.post20210130-py3.7.egg/odoo/fields.py", line 1120, in recompute
    self.compute_value(recs)
  File "/home/cirobox/odoo/odoo14/apps/odoo/lib/odoo-14.0.post20210130-py3.7.egg/odoo/fields.py", line 1142, in compute_value
    records._compute_field_value(self)
  File "/home/cirobox/odoo/odoo14/apps/odoo/lib/odoo-14.0.post20210130-py3.7.egg/odoo/addons/mail/models/mail_thread.py", line 412, in _compute_field_value
    return super()._compute_field_value(field)
  File "/home/cirobox/odoo/odoo14/apps/odoo/lib/odoo-14.0.post20210130-py3.7.egg/odoo/models.py", line 4059, in _compute_field_value
    getattr(self, field.compute)()
  File "/home/cirobox/odoo/odoo14/apps/odoo/lib/l10n-italy/l10n_it_withholding_tax/models/account.py", line 316, in _compute_amount_withholding_tax
    if not line.withholding_tax_generated_by_move_id:
Exception

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/cirobox/odoo/odoo14/apps/odoo/lib/odoo-14.0.post20210130-py3.7.egg/odoo/http.py", line 639, in _handle_exception
    return super(JsonRequest, self)._handle_exception(exception)
  File "/home/cirobox/odoo/odoo14/apps/odoo/lib/odoo-14.0.post20210130-py3.7.egg/odoo/http.py", line 315, in _handle_exception
    raise exception.with_traceback(None) from new_cause
AttributeError: 'NewId' object has no attribute 'withholding_tax_generated_by_move_id'`

Quando una nuova riga è in fase di creazione sembra che questo attributo non è disponibile.
Commentando da riga 314 a riga 316 l'esecuzione continua, ma si presenta un errore quando viene selezionata la ritenuta in riga.

**Comment by CiroBoxHub** [comment]
Buongiorno, in virtù di questo problema è della modifica effettuata [https://github.com/OCA/l10n-italy/pull/1984#discussion_r606246679](url) è necessario inserire anche il m2m delle ritenute nella form altrimenti non è possibile gestirle in riga.

**Comment by fredzamoabg** [comment]
@TheMule71 ho portato avanti la migrazione in https://github.com/TheMule71/l10n-italy/pull/9 (PR aperta direttamente sul tuo branch) puoi verificare?

**Comment by eLBati** [comment]
Anche: https://github.com/OCA/l10n-italy/issues/2223

**Comment by eLBati** [comment]
e https://github.com/OCA/l10n-italy/issues/2252

**Comment by TheMule71** [comment]
rebase alla ultima 14.0

**Comment by tafaRU** [comment]
Vedi https://github.com/OCA/l10n-italy/issues/2252

**Comment by TheMule71** [comment]
Nota: Travis fallisce ma qui da me...
```2021-05-08 16:01:45,683 10553 INFO odoo14-oca-wt odoo.modules.loading: Module l10n_it_withholding_tax: loading demo 
2021-05-08 16:01:45,728 10553 INFO odoo14-oca-wt odoo.addons.l10n_it_withholding_tax.tests.test_withholding_tax: Starting TestWithholdingTax.test_overlapping_rates ... 
2021-05-08 16:01:46,426 10553 INFO odoo14-oca-wt odoo.addons.l10n_it_withholding_tax.tests.test_withholding_tax: Starting TestWithholdingTax.test_partial_payment ... 
2021-05-08 16:01:47,429 10553 INFO odoo14-oca-wt odoo.addons.l10n_it_withholding_tax.tests.test_withholding_tax: Starting TestWithholdingTax.test_withholding_tax ... 
2021-05-08 16:01:48,360 10553 INFO odoo14-oca-wt odoo.modules.loading: Module l10n_it_withholding_tax loaded in 4.39s (incl. 2.60s test), 567 queries (+1150 test) 
```

su db nuovo.

**Comment by fredzamoabg** [comment]
> Nota: Travis fallisce ma qui da me...
> 
> ```
> 2021-05-08 16:01:45,728 10553 INFO odoo14-oca-wt odoo.addons.l10n_it_withholding_tax.tests.test_withholding_tax: Starting TestWithholdingTax.test_overlapping_rates ... 
> 2021-05-08 16:01:46,426 10553 INFO odoo14-oca-wt odoo.addons.l10n_it_withholding_tax.tests.test_withholding_tax: Starting TestWithholdingTax.test_partial_payment ... 
> 2021-05-08 16:01:47,429 10553 INFO odoo14-oca-wt odoo.addons.l10n_it_withholding_tax.tests.test_withholding_tax: Starting TestWithholdingTax.test_withholding_tax ... 
> 2021-05-08 16:01:48,360 10553 INFO odoo14-oca-wt odoo.modules.loading: Module l10n_it_withholding_tax loaded in 4.39s (incl. 2.60s test), 567 queries (+1150 test) 
> ```
> 
> su db nuovo.

se non erro fallisce un test del modulo **account** per https://github.com/OCA/l10n-italy/pull/1988/files#diff-d08c02c4bad4815b21a95522f70b3e16a9429c5ca1ff7df5707b88f60bd44486R57

**Comment by eLBati** [comment]
@TheMule71 se fai pulizia dei commit poi per me si può procedere

**Comment by TheMule71** [comment]
> @TheMule71 se fai pulizia dei commit poi per me si può procedere

Fatto.

**Review by eLBati — APPROVED** [review]

**Comment by eLBati** [comment]
@CiroBoxHub per te OK procedere?

**Review by robyf70 — COMMENTED** [review]
ho rimosso il commento perchè ho sbagliato

File: l10n_it_withholding_tax/views/account.xml [hunk 1]
```diff
@@ -213,20 +132,17 @@
...
-                    name="invoice_line_tax_wt_ids"
-                    widget="many2many_tags"
-                    options="{'no_create': True}"
-                    invisible="context.get('type', 'out_invoice') not in ['in_invoice', 'in_refund']"
-                />
+                        name="invoice_line_tax_wt_ids"
+                        widget="many2many_tags"
+                        options="{'no_create': True}"
+                    />
+                    <!-- attrs="{'invisible': context.get('move_type', 'out_invoice') not in ['in_invoice', 'in_refund']}" -->
```
Comment (tafaRU) [inline l10n_it_withholding_tax/views/account.xml:145]: come mai è stato commentato?

**Review by tafaRU — CHANGES_REQUESTED** [review]

File: l10n_it_withholding_tax/models/withholding_tax.py [hunk 2]
```diff
@@ -268,7 +268,7 @@ def get_wt_competence(self, amount_reconcile):
                     amount_wt = round(
                         base * wt_inv.tax_coeff, dp_obj.precision_get("Account")
                     )
-                if st.invoice_id.type in ["in_refund", "out_refund"]:
+                if st.invoice_id.move_type in ["in_refund", "out_refund"]:
```
[the remark this answers is not in the document]
Comment (TheMule71) [reply l10n_it_withholding_tax/models/withholding_tax.py:271]: non c'è in metodo già pronto nel model per le sole refund :(

File: l10n_it_withholding_tax/models/withholding_tax.py [hunk 3]
```diff
@@ -434,7 +428,8 @@ def generate_account_move(self):
 
         move_vals["line_ids"] = move_lines
         move = self.env["account.move"].create(move_vals)
-        move.post()
+        move.action_post()
```
[the remark this answers is not in the document]
Comment (TheMule71) [reply l10n_it_withholding_tax/models/withholding_tax.py:431]: La questione era ._post() or .action_post()? Visto che il modulo estende la seconda:
https://github.com/OCA/l10n-italy/pull/1988/files#diff-d08c02c4bad4815b21a95522f70b3e16a9429c5ca1ff7df5707b88f60bd44486R382
mi sembra più sensato invocare quella. Invocando la ._post() si salta il codice aggiunto da questo modulo.

File: l10n_it_withholding_tax/models/withholding_tax.py [hunk 4]
```diff
@@ -11,27 +11,28 @@ class WithholdingTax(models.Model):
...
-        if rate:
-            self.tax = rate[0]
-            self.base = rate[1]
-        else:
-            self.tax = 0
-            self.base = 1
+        for wt in self:
+            wt.tax = 0
+            wt.base = 1
+            if not wt._origin.id:
```
[the remark this answers is not in the document]
Comment (TheMule71) [reply l10n_it_withholding_tax/models/withholding_tax.py:21]: Fatto.

File: l10n_it_withholding_tax/tests/test_withholding_tax.py [hunk 5]
```diff
@@ -164,40 +153,40 @@ def test_withholding_tax(self):
...
-
-        # WT amount in payment move lines
-        self.assertTrue(
-            set(self.invoice.payment_move_line_ids.mapped("debit")) == {800, 200}
-        )
-
-        # WT aomunt applied in statement
+        register_payments.action_create_payments()
+
+        # # WT payment generation
```
[the remark this answers is not in the document]
Comment (TheMule71) [reply l10n_it_withholding_tax/tests/test_withholding_tax.py:158]: Sistemato, ho creato (spero) l'equivalmente in termini della 14.0.

File: l10n_it_withholding_tax/tests/test_withholding_tax.py [hunk 6]
```diff
@@ -210,14 +199,15 @@ def test_partial_payment(self):
...
             )
         )
-        register_payments.action_validate_invoice_payment()
+        register_payments.action_create_payments()
 
-        # WT amount in payment move lines
-        self.assertTrue(
-            set(self.invoice.payment_move_line_ids.mapped("debit")) == {600, 150}
-        )
+        # # WT amount in payment move lines
```
[the remark this answers is not in the document]
Comment (TheMule71) [reply l10n_it_withholding_tax/tests/test_withholding_tax.py:204]: idem :)

File: l10n_it_withholding_tax/models/withholding_tax.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (tafaRU) [reply l10n_it_withholding_tax/models/withholding_tax.py:21]: non vedo le modifiche, mi sai che non hai pusato, puoi verificare?

File: l10n_it_withholding_tax/views/account.xml (hunk 1 above)
Comment (TheMule71) [reply l10n_it_withholding_tax/views/account.xml]: Ripristinato come su 12.0 (pare che context.get() is un'eccezione alla regola) vd. Pedro

Comment (tafaRU) [reply l10n_it_withholding_tax/views/account.xml]: Vedi https://odoo-community.org/groups/contributors-15/contributors-196808?mode=date&date_begin=2021-05-01&date_end=2021-06-01

**Pull request merged**